A39-xds-http-filters.md

```diff
@@ -0,0 +1,268 @@
+- Every config in the list must be of a known type, as identified by the
+  map value proto message type.  Any unknown config type will cause the
+  resource to be NACKed, unless the `is_optional` field is true, in which
+  case the individual entry will be ignored.
+- The fields in the filter config should be validated by the filter
+  implementation.  Any validation error will cause the resource to be NACKed.
+- Note that gRPC will *not* fail validation if the map key specifies a
+  filter instance name that does not exist in the `HttpConnectionManager`
+  filter list.  This is because during an update, the xDS client code
+  cannot know which `HttpConnectionManager` config is currently being
+  used.
+- Similarly, gRPC will *not* fail validation if a map entry uses a
+  config for a filter that is supported on only the gRPC client but is
+  used on the gRPC server, or vice-versa.  This is because the xDS
+  client code cannot know whether a given RDS update will be used by a
+  gRPC client or server.
+
+Unlike Envoy, which provides an API that filters must call to access their
+per-{VirtualHost,Route,ClusterWeight} config overrides, gRPC will
+require filters to construct a merged configuration, applying any
+necessary overrides, when the xDS config is applied.  The filters will
+then use this merged configuration at run-time.  However, this approach
+is an implementation detail that could change in the future.
+
+Note that, for a given request, the filter's configuration will come
+from the top-level filter list in the `HttpConnectionManager` config
+and the most specific override.  In other words, if a given filter
+instance has an override in the `ClusterWeight` proto, that will be
+used; otherwise, if it has an override in the `Route` proto, that will
+be used; otherwise, if it has an override in the `VirtualHost` proto,
+that will be used; otherwise, no override will be used.
+
+### The Router Filter
+
+gRPC will support the [router
+filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/router_filter),
+which is used for the proto message type
+[`envoy.extensions.filters.http.router.v3.Router`](https://github.com/envoyproxy/envoy/blob/18db4c90e3295fb2c39bfc7b2ce641cfd6c3fbed/api/envoy/extensions/filters/http/router/v3/router.proto#L23).
+In practice, in order for gRPC to function properly, this filter is
+required and must be last in the filter list, although that is not
+enforced at config validation time.
+
+Note that gRPC will not actually have a discrete filter implementation
+for the router filter; instead, this filter will simply trigger the
+built-in routing functionality that we have already implemented in the
+resolver and LB policy plugins to the gRPC client channel.  However,
+in order to retain compatibility with Envoy, gRPC will still accept
+this filter configuration and treat it similarly to the way that Envoy
+does.
+
+Specifically, if the router filter is not present, gRPC will fail all
```

Comment:
In the discussion earlier today it sounded like we'd allow filters earlier in the list to still "have a go" at processing the request. But if the request gets to the end we'd fail it. That's a bit different than what's written here.

Reply:
Clarified.